client/src/connection/ssh/index.ts

```diff
@@ -233,6 +256,155 @@ export class SSHSession extends Session {
+
+  private promptForPassphrase = async (): Promise<string> => {
+    //TODO: need to think about whether these should be stored in secret storage
+    // I'm leaning towards no, but it's worth considering. Initial thought is that
+    // if users want to persist a passphrase, that the ssh-agent should be used,
+    // which seems to be inline with other solutions. Otherwise, the passphrase
+    // should be entered each time a session is established.
+    const passphrase = await window.showInputBox({
+      prompt: l10n.t("Enter the passphrase for the private key."),
+      password: true,
+    });
+    return passphrase;
+  };
+  private promptForPassword = async (): Promise<string> => {
+    //TODO: we need to properly manage the lifecycle of the password secret
+    const storedPassword = await this._secretStorage.get(this._passwordKey);
+    if (storedPassword) {
+      return storedPassword;
+    }
+
+    const source = new CancellationTokenSource();
+    this._cancellationSource = source;
+    const pw = await window.showInputBox(
+      {
+        ignoreFocusOut: true,
+        password: true,
+        prompt: l10n.t("Enter your password for this connection."),
+        title: l10n.t("Password Required"),
+      },
+      this._cancellationSource.token,
+    );
+    return pw;
+  };
+
+  private authHandler: AuthHandlerMiddleware = (
+    authsLeft: AuthenticationType[],
+    _partialSuccess: boolean,
+    cb: NextAuthHandler,
+  ) => {
+    if (!authsLeft) {
+      cb("none"); //sending none will usually prompt the server to send supported auth methods
+      return;
+    }
+
+    if (this._authMethods.length === 0) {
+      //if we're out of auth methods to try, then reject with an error
+      this._reject?.(
+        new Error(l10n.t("Could not authenticate to the SSH server.")),
+      );
+
+      //returning false will stop the auth process
+      return false;
+    }
+
+    //otherwise, fetch the next auth method to try
+    const authMethod = this._authMethods.shift();
+
+    //make sure the auth method is supported by the server
+    if (authsLeft.includes(authMethod)) {
+      switch (authMethod) {
+        //TODO: this is ugly and needs to be broken up into smaller functions
+        // one function per type of auth method would be a good start
+        case "publickey": {
+          //user set a keyfile path in profile config
+          //check for passphrase, prompt if necessary
+          //and then attempt to auth
+          if (this._config.privateKeyFilePath) {
+            //TODO: handle "file not found" errors
+            const keyContents = readFileSync(this._config.privateKeyFilePath);
+            const parsedKeyResult = utils.parseKey(keyContents);
+            // key is encrypted, prompt for passphrase
+            if (
+              parsedKeyResult instanceof Error &&
+              parsedKeyResult.message ===
+                "Encrypted OpenSSH private key detected, but no passphrase given"
+            ) {
```

Comment:
The ssh2 library has a function that can assist with detecting that the key file specified on the config property is encrypted, which is a nice help to allow users to specify shielded and unshielded keys on the session object (if they dont want to use agent). Right now with the code below if we detect a passphrase we'd prompt for it.